rust/datafusion/src/execution/physical_plan/expressions.rs

```diff
@@ -966,11 +966,32 @@ impl fmt::Display for BinaryExpr {
 
 impl PhysicalExpr for BinaryExpr {
     fn data_type(&self, input_schema: &Schema) -> Result<DataType> {
-        self.left.data_type(input_schema)
+        Ok(match self.op {
+            Operator::And
+            | Operator::Or
+            | Operator::Not
+            | Operator::NotLike
+            | Operator::Like
+            | Operator::Lt
+            | Operator::LtEq
+            | Operator::Eq
+            | Operator::NotEq
+            | Operator::Gt
+            | Operator::GtEq => DataType::Boolean,
+            Operator::Plus
+            | Operator::Minus
+            | Operator::Multiply
+            | Operator::Divide
+            | Operator::Modulus => {
+                // this assumes that the left and right expressions have already been co-coerced
+                // to the same type
+                self.left.data_type(input_schema)?
```

Comment:
I'm pretty sure this still isn't correct, but is closer than what we had before.

Reply:
Yes.

In another PR (now closed), what I did was to move this to a new function that accepts both a physical or logical expressions (a trait object `DataTyped`), and returns the correct type based on either of them. I.e.

```
fn binary_op_type(lhs: DataTyped, rhs: DataTyped, schema) -> Result<DataType>;
```

where `DataTyped` is a trait of a single method, `fn data_type(schema) -> Result<DataType>`. This way we do not have to keep track of these in 2 different places in the code.

Reply:
it might be worth a ticket to track the work that `data_type for BinaryExpr is incorrect